### ISSUE TITLE
patch for `curve25519-dalek` resolved to more than one candidate

### ISSUE THREAD

**Issue by olof-nord** [whisperfish/libsignal-service-rs#153]
Hello!

I am packaging the [axolotl](https://github.com/nanu-c/axolotl) project for Flathub.

The axolotl project uses the [crayfish](https://github.com/nanu-c/crayfish) library, which in turn depends on `libsignal-service` and `libsignal-service-hyper`.

In the main Cargo.toml in this project, a [patch](https://github.com/whisperfish/libsignal-service-rs/blob/master/Cargo.toml#L8) for crates-io is done, which specifies the `lizard2` branch for the `curve25519-dalek` library.

When using the flatpak cargo tooling, the below error message is shown.

From what I can tell, a possible solution would be to specify if this project prefers the `2.0.0` or the `3.2.0` version of the `curve25519-dalek` library. I could probably do this with some `sed` hacking, but before getting started I am curious which one of these versions is the right one to use.

Any input from your side to this? 
I am leaning into the 3.2.0 version, as higher numbers are always better, right? 

If I have misunderstood the error message, and this does not have anything to do with this library please let me know - it could be that this is an issue with the [cargo](https://github.com/flatpak/flatpak-builder-tools/tree/master/cargo) flatpak-builder-tools itself.

```toml
[patch.crates-io]
"curve25519-dalek" = { git = 'https://github.com/signalapp/curve25519-dalek', branch = 'lizard2', version = '3.2.0' }
```

```log
error: failed to resolve patches for `https://github.com/rust-lang/crates.io-index`

Caused by:
  patch for `curve25519-dalek` in `https://github.com/rust-lang/crates.io-index` failed to resolve

Caused by:
  patch for `curve25519-dalek` in `https://github.com/signalapp/curve25519-dalek?branch=lizard2` resolved to more than one candidate
  Found versions: 2.0.0, 3.2.0
  Update the patch definition to select only one package.
  For example, add an `=` version requirement to the patch definition, such as `version = "=3.2.0"`.
```

**Comment by rubdos** [comment]
Whisperfish:

```
❯ cargo tree | grep 25519
│   ├── curve25519-dalek v3.2.1 (https://github.com/signalapp/curve25519-dalek?branch=lizard2#4f0aa665)
│   └── x25519-dalek v1.2.0
│       ├── curve25519-dalek v3.2.1 (https://github.com/signalapp/curve25519-dalek?branch=lizard2#4f0aa665) (*)
│       ├── curve25519-dalek v3.2.1 (https://github.com/signalapp/curve25519-dalek?branch=lizard2#4f0aa665) (*)
│       │   ├── curve25519-dalek v3.2.1 (https://github.com/signalapp/curve25519-dalek?branch=lizard2#4f0aa665) (*)
```

So, as far as I know, we have no dependency on 2.0 at all any more. I'm quite sure I fixed and documented that in here: https://github.com/signalapp/curve25519-dalek/pull/3 and here: https://github.com/signalapp/curve25519-dalek/pull/4

Could you dump me the output of `cargo tree` here for axolotl? That command should make it clear what library actually depends on 2.0.0.

Also, note that the `patch` section you mention is one you *need* to also include in the main Rust build, because `patch` sections only get used at that level, as far as I know. This should probably be clarified in the README of this library.

**Comment by rubdos** [comment]
I updated the README in f3e2ceb7cecc2f0aead209509efa071dbd37ca37 to document this a bit. Feel free to suggest edits to it, in here or in a PR!

**Comment by olof-nord** [comment]
Many thanks for your assistance @rubdos! The readme update greatly clarifies this, in the case of `crayfish` the override was in place, without me understanding why.

The axolotl project is adding the `crayfish` as a git submodule. Messing around a bit with this, this error might be as simple as my local git submodule not being up to date. Either way, problem is solved!

**Comment by rubdos** [comment]
Great! Closing issue then as fixed by f3e2ceb.
